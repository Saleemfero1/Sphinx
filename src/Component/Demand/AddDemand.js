import React, { useState, useEffect, useContext } from "react";
import { useNavigate } from "react-router-dom";
import LocationServices from "../../Service/LocationServices";
import ItemServices from "../../Service/ItemServices";
import DemandServices from "../../Service/DemandServices";
import { ToastContainer, toast } from "react-toastify";
import { AuthContext } from "../../TokenDetails/AuthContext";
import { Button } from "@mui/material";
import Select from "react-select";

export default function AddDemand() {
  const { token } = useContext(AuthContext);
  const queryParameters = new URLSearchParams(window.location.search);
  const id = queryParameters.get("id");
  const [demandTypes] = useState(["PLANNED", "HARDPROMISED"]);
  const navigate = useNavigate();
  const [items, setItems] = useState([]);
  const [locations, setLocations] = useState([]);

  const [demand, setDemand] = useState({
    itemId: "",
    locationId: "",
    demandType: "",
    quantity: "",
  });

  useEffect(() => {
    if (id == -1) {
      return;
    } else {
      DemandServices.getDemandById(
        sessionStorage.getItem("organizationId"),
        id,
        token
      )
        .then((response) => {
          let existDemand = response.data;
          setDemand(existDemand);
        })
        .catch((err) => {
          console.log(err);
        });
    }
  }, []);

  useEffect(() => {
    ItemServices.getActiveItems(sessionStorage.getItem("organizationId"), token)
      .then((response) => {
        setItems(response.data);
      })
      .catch((err) => {
        console.log(err);
      });
  }, []);

  useEffect(() => {
    LocationServices.getLocations(
      sessionStorage.getItem("organizationId"),
      token
    )
      .then((response) => {
        setLocations(response.data);
      })
      .catch((err) => {
        console.log(err);
      });
  }, []);

  const optionItems = items.map((value) => ({
    label: value[Object.keys(value)[0]],
    value: value[Object.keys(value)[0]],
  }));

  const optionLocations = locations.map((value) => ({
    label: value[Object.keys(value)[0]],
    value: value[Object.keys(value)[0]],
  }));

  const saveDemand = (event) => {
    event.preventDefault();
    let newDemand = {
      itemId: demand.itemId,
      locationId: demand.locationId,
      demandType: demand.demandType,
      quantity: demand.quantity,
    };

    if (id == -1) {
      DemandServices.createDemand(
        sessionStorage.getItem("organizationId"),
        newDemand,
        sessionStorage.getItem("token")
      )
        .then((response) => {
          navigate("/demand", { state: { demandCreated: true } });
        })
        .catch((err) => {
          toast.error(err.response.data.message);
        });
    } else {
      DemandServices.updateDemand(
        sessionStorage.getItem("organizationId"),
        id,
        newDemand,
        token
      )
        .then((response) => {
          navigate("/demand", { state: { demandUpdated: true } });
        })
        .catch((err) => {
          toast.error(err.response.data.message);
        });
    }
  };

  const onChangeItemId = (itemId) => {
    setDemand({ ...demand, itemId: itemId.value });
  };

  const onChangeLocationId = (locationId) => {
    setDemand({ ...demand, locationId: locationId.value });
  };

  const onChangeDemandType = (event) => {
    setDemand({ ...demand, demandType: event.target.value });
  };

  const onChangeQuantity = (event) => {
    setDemand({ ...demand, quantity: event.target.value });
  };

  return (
    <div className="container mt-5">
      <ToastContainer position="bottom-left" />
      <div className="h3">
        {id == -1 ? "Add Demand" : "Upadte Demand"}
      </div>
      <form
        className="mt-3 p-3 shadow addItems"
        onSubmit={saveDemand}
        data-testid="add-demand-form"
      >
        <div className=" row justify-content-around ">
          <div className="form-group col-md-5 mb-3">
            <label htmlFor="itemId" data-testid="itemId">
              Select Item
            </label>
            <Select
              placeholder={id == -1 ? "Search item" : demand.itemId}
              options={optionItems}
              value={demand.itemId["itemId"]}
              onChange={onChangeItemId}
            ></Select>
          </div>
          <div className="form-group col-md-5 mb-3">
            <label htmlFor="LocationId" data-testid="locationId">
              Location Id
            </label>
            <Select
              placeholder={id == -1 ? "Serach location" : demand.locationId}
              options={optionLocations}
              value={demand.locationId["locationId"]}
              onChange={onChangeLocationId}
            ></Select>
          </div>
        </div>

        <div className="row justify-content-around">
          <div className="form-group col-md-5 mb-3">
            <label htmlFor="demandType" data-testid="demandType">
              Demand Type
            </label>
            <select
              className="form-select"
              aria-label="Default select example"
              value={demand.demandType}
              onChange={onChangeDemandType}
              required
            >
              <option value="">Choose Demand</option>
              {demandTypes.map((demand, index) => (
                <option key={demand}>{demand}</option>
              ))}
            </select>
          </div>

          <div className="form-group col-md-5 mb-3">
            <label htmlFor="Quantity" data-testid="Quantity">
              Quantity
            </label>
            <input
              type="text"
              className="form-control"
              id="Quantity"
              placeholder="Enter Quantity"
              value={demand.quantity}
              onChange={onChangeQuantity}
              required
            />
          </div>
        </div>
        <div className="row justify-content-around">
          <div className="col-md-5 mb-3">
            <Button
              variant="contained"
              className="bg-info text-black"
              type="submit"
              sx={{ mt: 2 }}
            >
              {id == -1 ? "Add Demand" : "Upadte Demand"}
            </Button>
            <Button
              variant="contained"
              color="error"
              sx={{ ml: 2, mt: 2 }}
              onClick={() => navigate("/demand")}
            >
              Cancel
            </Button>
          </div>
          <div className="col-md-5 mb-3">

          </div>
        </div>

      </form>
    </div>
  );
}
