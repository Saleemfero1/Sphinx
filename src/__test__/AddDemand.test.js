import React from "react";
import { render, screen } from "@testing-library/react";
import AddDemand from "../Component/Demand/AddDemand";
import { AuthProvider, AuthContext } from "../TokenDetails/AuthContext";
import { MemoryRouter } from "react-router-dom";

describe("test cases for Add Demand", () => {
  test("renders AddDemand component correctly", () => {
    render(
      <AuthProvider>
        <MemoryRouter>
          <AddDemand />
        </MemoryRouter>
      </AuthProvider>
    );

    expect(screen.getByTestId("itemId")).toBeInTheDocument();
    expect(screen.getByTestId("locationId")).toBeInTheDocument();
    expect(screen.getByTestId("demandType")).toBeInTheDocument();
    expect(screen.getByTestId("Quantity")).toBeInTheDocument();
    expect(screen.getByTestId("add-demand-form")).toMatchSnapshot();
  });

  test("test firEvent for Inputs", () => {
    render(
      <AuthProvider>
        <MemoryRouter>
          <AddDemand />
        </MemoryRouter>
      </AuthProvider>
    );
  });
});
