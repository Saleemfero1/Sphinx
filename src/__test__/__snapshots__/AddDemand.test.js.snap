// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test cases for Add Demand renders AddDemand component correctly 1`] = `
<form
  class="mt-3 p-3 shadow addItems"
  data-testid="add-demand-form"
>
  <div
    class=" row mb-3 "
  >
    <div
      class="form-group col-md-6"
    >
      <label
        data-testid="itemId"
        for="itemId"
      >
        Select Item
      </label>
      <div
        class=" css-b62m3t-container"
      >
        <span
          class="css-1f43avz-a11yText-A11yText"
          id="react-select-2-live-region"
        />
        <span
          aria-atomic="false"
          aria-live="polite"
          aria-relevant="additions text"
          class="css-1f43avz-a11yText-A11yText"
        />
        <div
          class=" css-13cymwt-control"
        >
          <div
            class=" css-16674sh-Component"
          >
            <div
              class=" css-1jqq78o-placeholder"
              id="react-select-2-placeholder"
            />
            <div
              class=" css-d9r47t-Component"
              data-value=""
            >
              <input
                aria-autocomplete="list"
                aria-describedby="react-select-2-placeholder"
                aria-expanded="false"
                aria-haspopup="true"
                autocapitalize="none"
                autocomplete="off"
                autocorrect="off"
                class=""
                id="react-select-2-input"
                role="combobox"
                spellcheck="false"
                style="opacity: 1; width: 100%; grid-area: 1 / 2; min-width: 2px; border: 0px; margin: 0px; outline: 0; padding: 0px;"
                tabindex="0"
                type="text"
                value=""
              />
            </div>
          </div>
          <div
            class=" css-ny0e4k-Component"
          >
            <span
              class=" css-1u9des2-indicatorSeparator"
            />
            <div
              aria-hidden="true"
              class=" css-1xc3v61-indicatorContainer"
            >
              <svg
                aria-hidden="true"
                class="css-tj5bde-Svg"
                focusable="false"
                height="20"
                viewBox="0 0 20 20"
                width="20"
              >
                <path
                  d="M4.516 7.548c0.436-0.446 1.043-0.481 1.576 0l3.908 3.747 3.908-3.747c0.533-0.481 1.141-0.446 1.574 0 0.436 0.445 0.408 1.197 0 1.615-0.406 0.418-4.695 4.502-4.695 4.502-0.217 0.223-0.502 0.335-0.787 0.335s-0.57-0.112-0.789-0.335c0 0-4.287-4.084-4.695-4.502s-0.436-1.17 0-1.615z"
                />
              </svg>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      class="form-group col-md-6"
    >
      <label
        data-testid="locationId"
        for="LocationId"
      >
        Location Id
      </label>
      <div
        class=" css-b62m3t-container"
      >
        <span
          class="css-1f43avz-a11yText-A11yText"
          id="react-select-3-live-region"
        />
        <span
          aria-atomic="false"
          aria-live="polite"
          aria-relevant="additions text"
          class="css-1f43avz-a11yText-A11yText"
        />
        <div
          class=" css-13cymwt-control"
        >
          <div
            class=" css-16674sh-Component"
          >
            <div
              class=" css-1jqq78o-placeholder"
              id="react-select-3-placeholder"
            />
            <div
              class=" css-d9r47t-Component"
              data-value=""
            >
              <input
                aria-autocomplete="list"
                aria-describedby="react-select-3-placeholder"
                aria-expanded="false"
                aria-haspopup="true"
                autocapitalize="none"
                autocomplete="off"
                autocorrect="off"
                class=""
                id="react-select-3-input"
                role="combobox"
                spellcheck="false"
                style="opacity: 1; width: 100%; grid-area: 1 / 2; min-width: 2px; border: 0px; margin: 0px; outline: 0; padding: 0px;"
                tabindex="0"
                type="text"
                value=""
              />
            </div>
          </div>
          <div
            class=" css-ny0e4k-Component"
          >
            <span
              class=" css-1u9des2-indicatorSeparator"
            />
            <div
              aria-hidden="true"
              class=" css-1xc3v61-indicatorContainer"
            >
              <svg
                aria-hidden="true"
                class="css-tj5bde-Svg"
                focusable="false"
                height="20"
                viewBox="0 0 20 20"
                width="20"
              >
                <path
                  d="M4.516 7.548c0.436-0.446 1.043-0.481 1.576 0l3.908 3.747 3.908-3.747c0.533-0.481 1.141-0.446 1.574 0 0.436 0.445 0.408 1.197 0 1.615-0.406 0.418-4.695 4.502-4.695 4.502-0.217 0.223-0.502 0.335-0.787 0.335s-0.57-0.112-0.789-0.335c0 0-4.287-4.084-4.695-4.502s-0.436-1.17 0-1.615z"
                />
              </svg>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
  <div
    class=" row"
  >
    <div
      class="form-group col-md-6"
    >
      <label
        data-testid="demandType"
        for="demandType"
      >
        Demand Type
      </label>
      <select
        aria-label="Default select example"
        class="form-select"
        required=""
      >
        <option
          value=""
        >
          Choose Demand
        </option>
        <option>
          PLANNED
        </option>
        <option>
          HARDPROMISED
        </option>
      </select>
    </div>
    <div
      class="form-group col-md-6"
    >
      <label
        data-testid="Quantity"
        for="Quantity"
      >
        Quantity
      </label>
      <input
        class="form-control"
        id="Quantity"
        placeholder="Enter Quantity"
        required=""
        type="text"
        value=""
      />
    </div>
  </div>
  <button
    class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium bg-info text-black css-lll1vm-MuiButtonBase-root-MuiButton-root"
    tabindex="0"
    type="submit"
  >
    Upadte Demand
    <span
      class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
    />
  </button>
  <button
    class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedError MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-root MuiButton-contained MuiButton-containedError MuiButton-sizeMedium MuiButton-containedSizeMedium css-7ba66y-MuiButtonBase-root-MuiButton-root"
    tabindex="0"
    type="button"
  >
    Cancel
    <span
      class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
    />
  </button>
</form>
`;
